### PR TITLE
Upgrade asm to 6.1

### LINKED ISSUE #4695: jdk10 compat: Upgrade asm to 6.1
Asm 6.1 has been released, and it has finally support for jdk10 (notably with the addition of `org.objectweb.asm.Opcodes.V10`).  Groovy has already updated asm to 6.1, and it will be in the 2.5.0-rc-1 release which is due any day now.

### PR DESCRIPTION
**Pull request by henrik242**
Improves jdk 10 compatibility.  Fixes #4695

### CODE REVIEW

**Comment by blindpirate** [comment]
Thanks @henrik242 !

File: subprojects/distributions/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy [hunk 1]
```diff
@@ -38,7 +38,7 @@ abstract class DistributionIntegrationSpec extends AbstractIntegrationSpec {
     abstract String getDistributionLabel()
 
     int getLibJarsCount() {
-        192
+        193
```
Comment (oehme) [inline subprojects/distributions/src/integTest/groovy/org/gradle/DistributionIntegrationSpec.groovy]: ❌ Why did the library count increase?

**Comment by blindpirate** [comment]
@oehme Thanks for pointing out this. 

ASM 6.0 has 4 jars: `asm-commons-6.0.jar`/`asm-tree-6.0.jar`/`asm-6.0.jar`/`asm-util-6.0.jar` and ASM 6.1 has 5: `asm-commons-6.1.jar`/`asm-tree-6.1.jar`/`asm-6.1.jar`/`asm-util-6.1.jar`/`asm-analysis-6.1.jar` so I guess it's necessary here.

I'll add @henrik242 to contributor section of release note soon.

**Comment by bmuschko** [comment]
Optimally, this would make the changes in [`PatchedClassReader`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/util/internal/PatchedClassReader.java#L33) unnecessary or we should change the handling to cover Java 11 instead.

**Comment by Godin** [comment]
Could be noted that ASM 6.1 has following regression https://gitlab.ow2.org/asm/asm/issues/317815
Hope this helps.

**Comment by Godin** [comment]
Also about increase of JARs could be noted that asm-analysis JAR already existed in 6.0, and in 6.1 dependency on it was added to asm-utils.

**Comment by Godin** [comment]
And about

>  change the handling to cover Java 11 instead

could be noted that [JEP 309](http://openjdk.java.net/jeps/309) is scheduled on JDK 11, so that JDK 11 supposed to have a real change in bytecode and not just increment of version. And there is chance of  other changes in bytecode - for example [JEP 181](http://openjdk.java.net/jeps/181) currently is in candidate state, but AFAIK some time ago was scheduled on 11 and earlier even on 10. Thus trick with downgrade might not work reliably in all cases.

**Comment by blindpirate** [comment]
@Godin Thanks a lot for the information! We'll reconsider this change.

**Comment by oehme** [comment]
The issue has been resolved, we should ask the asm team for a patch release this week or early next week so we can release with Java 10 compatibility.

**Comment by eriwen** [comment]
@oehme @blindpirate I submitted https://github.com/gradle/gradle/pull/4852

**Pull request merged**